### ISSUE TITLE
How to configure high-performance logging into MongoDB

### ISSUE THREAD

**Issue by Wernfried** [syslog-ng/syslog-ng#3454]
I have an application which receives log messages from a firewall. The logs are written into a MongoDB. My goal is to process 30'000 messages per second (more or less constantly for 7*24 hours, not as transient peak value)
As peak value I expect app. 50'000 messages per second.

With several settings I reached up to 20'000 msg/sec. but that is not sufficient for our life traffic. The MongoDB host reach a load of app. 50% so I guess there is not the bottleneck. When I run my tests then the syslog-ng "memory_usage" and "queued" stat counters goes constantly up, so looks like the limit is rather on syslog-ng.

Can someone support me to configure syslog-ng properly? I really get stuck with the parameters, esp. I lost track which parameter is replaced by which new parameter.

My current configuration looks like this: At first stage the log messages are received by syslog-ng and then forwarded to 20 other syslog-ng processes in order to have parallel processing:

First stage config:

```
source s_loggen_hw { 
   network( 
      transport("udp"), 
      port(2690),
      so-rcvbuf(1000MiB),
      log-fetch-limit(100),
      log-iw-size(1000),
      flags(no-parse),
   ); 
}; 

filter f_mongo_1 { "${MONGO_CHUNK}" eq "1" };
filter f_mongo_2 { "${MONGO_CHUNK}" eq "2" };
...
filter f_mongo_20 { "${MONGO_CHUNK}" eq "20" }; 


destination d_dispatcher_1m {
   network("localhost",
      transport("tcp"),
      port(27648),
      flags(syslog-protocol),
      log-fifo-size(10kB),
      so-rcvbuf(1000MiB),
      so-sndbuf(1000MiB)
   );
};
destination d_dispatcher_2m {
   network("localhost",
      transport("tcp"),
      port(27649),
      flags(syslog-protocol),
      log-fifo-size(10kB),
      so-rcvbuf(1000MiB),
      so-sndbuf(1000MiB)
   );
}; 
...
destination d_dispatcher_20m {
   network("localhost",
      transport("tcp"),
      port(27667),
      flags(syslog-protocol),
      log-fifo-size(10kB),
      so-rcvbuf(1000MiB),
      so-sndbuf(1000MiB)
   );
}; 


log { 
   source(s_loggen_hw); 
   rewrite { set("$(+ $(% ${RCPTID} 20) 1)", value("MONGO_CHUNK") ); };
   log {
      filter(f_mongo_1);
      destination(d_dispatcher_1m);
      flags(final); 
   };
   log {
      filter(f_mongo_2);
      destination(d_dispatcher_2m);
      flags(final); 
   };
   ...
   log {
      filter(f_mongo_20);
      destination(d_dispatcher_20m);
      flags(final); 
   };
   flags(flow-control);
};
```

Then I have 20 almost equal second stages. Of course I tried different figures but the performance difference are just minor.

Second stage config:
```
source s_dispatcher { 
   network( 
      transport("tcp"),
      port(27648), 
      flags(syslog-protocol, validate-utf8),
      so-rcvbuf(1000MiB),
      log-fetch-limit(10kB),
      log-iw-size(10kB),
      max-connections(1)
   );
};

destination d_mongo_snat {
   mongodb(
      uri('mongodb://username:********@localhost:27100/mip?authSource=admin')
      collection("sessions")
      persist-name("snat")
      batch-timeout(1000)
      batch-lines(100000)
      disk-buffer(
         reliable(no)
         dir("/var/lib/mediation/buffer")
         disk-buf-size(62882MiB)
         mem-buf-length(1000kB)
         qout-size(497kB)
      )
      value-pairs(
         pair("r", boolean("1"))
         pair("v", boolean("${VALID}"))
         pair("t0", datetime("${BEGINTIME}${.HW.BeginTime}"))
         pair("t", datetime("${.FN.eventtime}${.HW.EndTime}"))
         pair("si", "${.FN.srcip}${.HW.SourceIP}")
         pair("sp", "${.FN.srcport}${.HW.SourcePort}")
         pair("tsi", "${.FN.transip}${.HW.SourceNatIP}")
         pair("tsp", int32("${.FN.transport}${.HW.SourceNatPort}"))
         pair("n", "${.FN.trandisp}${TRANDISP}")
         pair("di", "${.FN.dstip}${.HW.DestinationIP}")
         pair("dp", "${.FN.dstport}${.HW.DestinationPort}")
         pair("p", "${PROTOCOL}")
         pair("a", "${.FN.action}${ACTION}")
         pair("f", "${.FN.devname}${.HW.devname}")
         pair("sb", int64("0${.FN.sentbyte}${.HW.SendBytes}"))
         pair("rb", int64("0${.FN.rcvdbyte}${.HW.RcvBytes}"))
      )
   );
};

destination d_mongo_dnat {
   # for dnat the amount of messages is 10'000 times lower than snat
   mongodb(
      uri('mongodb://username:********@localhost:27100/mip?authSource=admin')
      collection("sessions")
      persist-name("dnat")
      batch-timeout(1000)
      batch-lines(10)
      disk-buffer(
         reliable(no)
         dir("/var/lib/mediation/buffer")
         disk-buf-size(6439116KiB)
         mem-buf-length(100000)
         qout-size(49700)
      )
      value-pairs(
         pair("r", boolean("1"))
         pair("v", boolean("${VALID}"))
         pair("t0", datetime("${BEGINTIME}${.HW.BeginTime}"))
         pair("t", datetime("${.FN.eventtime}${.HW.EndTime}"))
         pair("si", "${.FN.srcip}${.HW.SourceIP}")
         pair("sp", "${.FN.srcport}${.HW.SourcePort}")
         pair("tsi", boolean("0")) # required for sharding, resp. $merge
         pair("tdi", "${.FN.tranip}${.HW.DestinationNatIP}")
         pair("tdp", int32("${.FN.tranport}${.HW.DestinationNatPort}"))
         pair("n", "${.FN.trandisp}${TRANDISP}")
         pair("di", "${.FN.dstip}${.HW.DestinationIP}")
         pair("dp", "${.FN.dstport}${.HW.DestinationPort}")
         pair("p", "${PROTOCOL}")
         pair("a", "${.FN.action}${ACTION}")
         pair("f", "${.FN.devname}${.HW.devname}")
         pair("sb", int64("0${.FN.sentbyte}${.HW.SendBytes}"))
         pair("rb", int64("0${.FN.rcvdbyte}${.HW.RcvBytes}"))
      )
   );
};


log { 
   source(s_dispatcher);
   parser { kv-parser(prefix(".HW."), pair-separator(",")); };
   # some rewrite and filter stuff
   log {
      filter(f_huawei_valid);
      log { 
         filter(f_snat); 
         destination(d_mongo_snat); 
         flags(final); 
      };
      log { 
         filter(f_dnat); 
         destination(d_mongo_dnat); 
         flags(final); 
      };
      flags(final); 
   };
   log { 
      destination(d_firewall_other); 
      flags(final); 
   }; 
};

```

My machine has these figures:

```
$ df -h 
Filesystem                  Size  Used Avail Use% Mounted on
/dev/mapper/vg01-mediation  1.5T   13G  1.5T   1% /var/lib/mediation

echo $(( `cat /proc/meminfo | grep MemTotal | egrep -o [[:digit:]]+` /1024/1024 ))GiB
125GiB

$ cat /proc/cpuinfo | grep "model name" | head -1
model name      : Intel(R) Xeon(R) Gold 6132 CPU @ 2.60GHz

$ cat /proc/cpuinfo | grep "model name" | wc -l
28


$ syslog-ng -V
syslog-ng 3 (3.29.1)
Config version: 3.29
Installer-Version: 3.29.1
Revision: 
Compile-Date: Aug 29 2020 08:27:16
Module-Directory: /usr/lib64/syslog-ng
Module-Path: /usr/lib64/syslog-ng
Include-Path: /usr/share/syslog-ng/include
Available-Modules: add-contextual-data,affile,afprog,afsocket,afstomp,afuser,appmodel,basicfuncs,cef,confgen,cryptofuncs,csvparser,dbparser,disk-buffer,examples,graphite,hook-commands,json-plugin,kvformat,linux-kmsg-format,map-value-pairs,pseudofile,sdjournal,stardate,syslogformat,system-source,tags-parser,tfgetent,timestamp,xml,afsql,azure-auth-header,http,afsmtp,afmongodb
Enable-Debug: off
Enable-GProf: off
Enable-Memtrace: off
Enable-IPv6: on
Enable-Spoof-Source: on
Enable-TCP-Wrapper: on
Enable-Linux-Caps: on
Enable-Systemd: on

```

How to set different buffer and windows sizes for such a load?

Thanks for any suggestions.
Wernfried

**Comment by Wernfried** [comment]
$RCPTID works fine, the messages are distributed very evenly.  This first stage does not limit my application, it works fine up to 200k Msg/sec. without filling any queues or buffers. But I will remove the rewrite, thanks for the hint.

All 21 syslog-ng processes are running on one host. Of course, I could define all log() sections in a single syslog-ng instance but the performance is much worse. Multithreading is not sufficient. Actually I get performance saturation at around 40+1 syslog-ng processes.

Next trial from my side is to increase the number of mongos instances, i.e. one mongos instance for each syslog-ng process. Up to now I used only 10 of them for 20 syslog-ng processes, i.e. each mongos instance serves 2 syslog-ng processes.

Yes, I put qout-size() to a high value, I have 128GiB RAM which is plenty of space to buffer log messages. However, this is mainly used in case the MongoDB is temporarily not available. Typically this buffer should be more or less empty while normal operation. I am not concerned about a few thousands messages in this buffer but it is growing and growing and after some hours syslog-ng starts dropping the messages because the buffer (RAM + disc) becomes full.

**Comment by gaborznagy** [comment]
Even though you mentioned $RCPTID works fine, I would still recommend that you should loadbalance with $R_MSEC instead for performance reasons.

**Comment by gaborznagy** [comment]
Another question: I don't have any performance measure results for mongodb, so I don't know what is the maximum output rate that could be expected from mongodb destination. 
Did you measure with a minimal setup what was the highest possible output rate for a single mongodb destination?

**Comment by Wernfried** [comment]
As far as I remember I got around 2000 messages/sec when I did my first test with single destination. Perhaps I will try to setup an in-Memory MongoDB which is much faster and where is should be possible to determine the max. output rate for syslog-ng destination.

**Comment by gaborznagy** [comment]
Just checked the code and mongodb() destination doesn't use batches, messages are inserted one-by-one, so your `batch-timeout()` and `batch-lines()` config options for mongodb() doesn't work.

I see there is a bulk API in the mongoc library:
http://mongoc.org/libmongoc/current/bulk.html
I'll see if next week I can take a look into this.

**Comment by Wernfried** [comment]
Yes, I was suspecting that, because changing `batch-timeout()` and `batch-lines()` did not change anything. However, according documentation it is supported: https://www.syslog-ng.com/technical-documents/doc/syslog-ng-open-source-edition/3.26/administration-guide/42#kanchor3109

The Bulk operation would be definitely the better one.

**Comment by Wernfried** [comment]
Today I run again a test on a rather small machine (virtual, 4 CPU's, 8GB RAM). I configured a single mongo destination. 
Max. output was 900 messages/sec. If I put more than the memory-buffer fills up. In the past I never tested with only a single destination.

**Just for comparison:** 
When I use the Mongo native tools ([mongoimport](https://docs.mongodb.com/database-tools/mongoimport/)) then it takes about 3 seconds to insert 100k messages from a local file, i.e. around 30k messages/sec. I run it on the same machine with "numInsertionWorkers=1". When I increase the number of InsertionWorkers then 100k messages are inserted in much less than one second.

I also tested with an in-Memory MongoDB, i.e. the entire MongoDB is stored only in the RAM. With an im-Memory DB I reached up to 1800 messages/sec. (also with a single mongo destination). 

So, it really looks that syslog-ng is the bottleneck here, the MongoDB could easily handle by order of magnitude more data.  
If syslog-ng wants to keep the promise of a "high-performance log daemon" then Bulk operation for the mongo driver would be essential.

Best Regards
Wernfried

**Comment by gaborznagy** [comment]
Thanks for the feedback on the benchmarks you made!
Unfortunately, we had quite much on our plate to take an experiment with the bulk API yet, but I'm keeping this on my radar.

**Comment by Wernfried** [comment]
I run again some performance tests on my big machine (24 CPU's, 128 GB RAM)

With the `mongodb` driver I get up to 20k messages per second. Due to the lack of Mongo Bulk API I developed this workaround:


```
destination d_mongo_snat {
   program("/opt/mediation/mongo-insertMany.sh"
      flush-lines(10000)
      persist-name("snat")
      disk-buffer(
         reliable(no)
         dir("/var/lib/mediation/buffer")
         disk-buf-size(41921MiB)
         mem-buf-length(100kB)
         qout-size(1875kB)
      )
      template("{\"t0\":{\"$$date\":{\"$$numberLong\":\"${BEGINTIME}${.HW.BeginTime}000\"}}, \"di\":\"${.FN.dstip}${.HW.DestinationIP}\" ... }\n")
   );
};
```

The `/opt/mediation/mongo-insertMany.sh` script looks similar to this one:

```
#!/bin/bash

docs=()
import(){
   [[ ${#docs[@]} -gt 0 ]] && /usr/bin/mongoimport --db=mip --collection=sessions.raw --jsonArray <<< "[${docs[@]}]"
   docs=()
}

trap import USR1 SIGHUP SIGINT SIGTERM
while read line ; do
   docs+=( "$line" ) 
   [[ ${#docs[@]} -ge 10000 ]] && import
done
```

This works pretty well. 

Similar to the `mongod` driver I reach around 20k messages per second. 

With the `mongod` I have to configure massive parallel processes. I configured 50 or even 100 syslog-ng processes and 20+ mongos instances. The CPU load was around 95% on all 24 CPU's.

With the `program` driver I get 20k msg./sec. with just 10 syslog-ng processes and a single mongos instance. The CPU load is only a few percent!

So, the `program` driver works much more efficient than the `mongod` driver. However, I am not able to get more than 20k messages per second, no matter what I tried. Even when the logs are not processed like this:

```
#!/bin/bash
while read line ; do
   docs+=( "$line" ) 
   [[ ${#docs[@]} -ge 10000 ]] && docs=()
done
```

I don't get a rate higher than 20-22k msg/sec. What could be the limiting factor here? Any idea to find out (and of course how to get rid of it)?

**Comment by gaborznagy** [comment]
Hi, sorry for not replying. I've checked program dest quickly with the above script with strace and what I saw was that read() syscall reads 1 byte sizes from stdin.
I've tried with -n and -N options (latter strictly returns after reading nchar bytes) to increase the number of bytes read with each read syscall, but without any success. With -N I see 128 bytes long reads, but -N wouldn't be good for your case as it splits the log messages.

**Comment by Wernfried** [comment]
Hi

Actually I found a solution, I use `mongoimport` directly instead of bash script, i.e. 

```
destination d_mongo_snat {
   program("/usr/bin/mongoimport --numInsertionWorkers=10 --db=mip --collection=sessions.raw"
      ... 
      template("{ ... }\n")
   );
};
```

I configured 5 syslog-ng processes, with this I insert up to 120'000 messages per second! The CPU and memory load is still low with that.
It would be able to insert even more data, but at this rate I hit other limits of my system.

So, having the mongo Bulk API developed would still be a nice feature but with this solution it is not urgent anymore.

Thank you for your support
Wernfried

**Comment by gaborznagy** [comment]
Thanks for the feedback, I'm glad you've found a workaround!
We definitely keep the bulk API feature request open!

**Comment by MrAnno** [comment]
FYI #3621 adds multi-threaded connection pooling support for the MongoDB destination, which takes us one step further in terms of high-performance mongo logging.

**Comment by travisbell** [comment]
The multi threaded and worker enhancements are awesome but boy, would it still be great to see real batch support. Seems like a lot of the remote destinations support batching but for some reason MongoDB is left out. I hope you guys get to the final step and add batching. 😄

**Comment by Madhusmeeta** [comment]
I'd want to work on this issue. Please give this problem to me.

**Comment by Gauravjadhav22** [comment]
I wanna work on this assign me...

**Comment by MrAnno** [comment]
Resolved by #4483
